File: python-client/audio_streamer.py
#!/usr/bin/env python3
"""
Módulo para streaming de audio bidireccional
"""
import grpc
import pyaudio
import threading
import time
from proto import chat_pb2


class AudioStreamer:
    """
    Maneja el streaming de audio bidireccional con el servidor gRPC
    """

    # Configuración de audio
    AUDIO_FORMAT = pyaudio.paInt16
    CHANNELS = 1
    RATE = 44100
    CHUNK = 1024

    def __init__(self, stub, sender, room_id):
        """
        Inicializa el AudioStreamer

        Args:
            stub: Cliente gRPC
            sender: Nombre del usuario
            room_id: ID de la sala
        """
        self.stub = stub
        self.sender = sender
        self.room_id = room_id

        # Audio components
        self.audio = pyaudio.PyAudio()
        self.microphone_stream = None
        self.speakers_stream = None

        # Control de estados
        self.grpc_stream_active = False
        self.mic_active = False
        self.speakers_active = False

        # Thread y observer para gRPC
        self.request_observer = None
        self.mic_thread = None
        self.receive_thread = None
        self.response_iterator = None

    def _print_message(self, message):
        """
        Imprime un mensaje y redibuja el prompt
        """
        ansi_clear_line = "\r\x1b[2K"
        prompt = f"[{time.strftime('%H:%M')}] Tú: "
        print(f"{ansi_clear_line}{message}\n{prompt}", end="", flush=True)

    def start_audio_connection(self):
        """
        Establece la conexión bidireccional gRPC para audio
        """
        if self.grpc_stream_active:
            self._print_message("La conexión de audio gRPC ya está activa.")
            return

        # Crear metadatos para enviar sender y room_id
        metadata = (
            ('sender', self.sender),
            ('room-id', self.room_id),
        )

        try:
            # Iniciar el stream bidireccional con metadatos
            self.response_iterator = self.stub.StreamAudio(
                self._generate_audio_chunks(),
                metadata=metadata
            )

            self.grpc_stream_active = True

            # Iniciar thread para recibir audio
            self.receive_thread = threading.Thread(
                target=self._receive_audio,
                daemon=True
            )
            self.receive_thread.start()

            self._print_message("Conexión de audio gRPC establecida.")

        except grpc.RpcError as e:
            print(f"\nError al establecer conexión de audio: {e.details()}")
            self.grpc_stream_active = False

    def _generate_audio_chunks(self):
        """
        Generador que produce chunks de audio del micrófono
        """
        while self.grpc_stream_active:
            if self.mic_active and self.microphone_stream:
                try:
                    data = self.microphone_stream.read(self.CHUNK, exception_on_overflow=False)
                    yield chat_pb2.AudioChunk(data=data)
                except Exception as e:
                    print(f"\nError al capturar audio: {e}")
                    break
            else:
                # Si el mic no está activo, esperar un poco
                time.sleep(0.1)

    def _receive_audio(self):
        """
        Recibe audio del servidor y lo reproduce
        """
        try:
            for audio_chunk in self.response_iterator:
                if self.speakers_active and self.speakers_stream:
                    self.speakers_stream.write(audio_chunk.data)
        except grpc.RpcError as e:
            if e.code() not in (grpc.StatusCode.CANCELLED, grpc.StatusCode.UNAVAILABLE):
                print(f"\nError en el stream de audio gRPC: {e.details()}")
        except Exception as e:
            print(f"\nError al recibir audio: {e}")
        finally:
            self._print_message("Recepción de audio gRPC finalizada.")

    def stop_audio_connection(self):
        """
        Cierra la conexión bidireccional gRPC para audio
        """
        if not self.grpc_stream_active:
            return

        self.grpc_stream_active = False
        self.stop_mic()
        self.stop_speakers()

        # Esperar a que los threads terminen
        if self.receive_thread and self.receive_thread.is_alive():
            self.receive_thread.join(timeout=1)

        self._print_message("Conexión de audio gRPC cerrada.")

    def start_mic(self):
        """
        Inicia la captura de audio del micrófono
        """
        if self.mic_active:
            self._print_message("Micrófono ya activo.")
            return

        if not self.grpc_stream_active:
            print("\nPrimero debes establecer la conexión gRPC de audio (/mic on o /listen on).")
            return

        try:
            self.microphone_stream = self.audio.open(
                format=self.AUDIO_FORMAT,
                channels=self.CHANNELS,
                rate=self.RATE,
                input=True,
                frames_per_buffer=self.CHUNK
            )
            self.mic_active = True
            self._print_message("Micrófono activado. Transmitiendo voz...")

        except Exception as e:
            print(f"\nError al acceder al micrófono: {e}")
            self.mic_active = False

    def stop_mic(self):
        """
        Detiene la captura de audio del micrófono
        """
        if not self.mic_active:
            return

        self.mic_active = False

        if self.microphone_stream:
            self.microphone_stream.stop_stream()
            self.microphone_stream.close()
            self.microphone_stream = None

        self._print_message("Micrófono detenido.")

    def start_speakers(self):
        """
        Inicia la reproducción de audio por los altavoces
        """
        if self.speakers_active:
            self._print_message("Altavoces ya activos para reproducción.")
            return

        if not self.grpc_stream_active:
            print("\nPrimero debes establecer la conexión gRPC de audio (/mic on o /listen on).")
            return

        try:
            self.speakers_stream = self.audio.open(
                format=self.AUDIO_FORMAT,
                channels=self.CHANNELS,
                rate=self.RATE,
                output=True,
                frames_per_buffer=self.CHUNK
            )
            self.speakers_active = True
            self._print_message("Altavoces activados para reproducción de audio.")

        except Exception as e:
            print(f"\nError al acceder a los altavoces: {e}")
            self.speakers_active = False

    def stop_speakers(self):
        """
        Detiene la reproducción de audio por los altavoces
        """
        if not self.speakers_active:
            return

        self.speakers_active = False

        if self.speakers_stream:
            self.speakers_stream.stop_stream()
            self.speakers_stream.close()
            self.speakers_stream = None

        self._print_message("Altavoces detenidos.")

    def is_mic_active(self):
        """Retorna si el micrófono está activo"""
        return self.mic_active

    def is_speakers_active(self):
        """Retorna si los altavoces están activos"""
        return self.speakers_active

    def is_grpc_stream_active(self):
        """Retorna si el stream gRPC está activo"""
        return self.grpc_stream_active

    def cleanup(self):
        """
        Limpia recursos de audio
        """
        self.stop_audio_connection()
        self.audio.terminate()

File: java-client/src/main/java/com/chat/client/AudioStreamer.java
package com.chat.client;

import com.chat.grpc.AudioChunk;
import com.chat.grpc.ChatServiceGrpc;
import com.google.protobuf.ByteString;
import io.grpc.CallOptions;
import io.grpc.ClientCall;
import io.grpc.ClientInterceptor;
import io.grpc.ForwardingClientCall;
import io.grpc.ForwardingClientCallListener;
import io.grpc.Metadata;
import io.grpc.MethodDescriptor;
import io.grpc.Channel;
import io.grpc.stub.StreamObserver;

import javax.sound.sampled.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.io.IOException;

public class AudioStreamer {

    private static final DateTimeFormatter TIME_FORMATTER = DateTimeFormatter.ofPattern("HH:mm");

    private final ChatServiceGrpc.ChatServiceStub asyncStub;
    private final String sender;
    private final String roomId;

    // Componentes de audio
    private AudioFormat audioFormat;
    private TargetDataLine microphone;
    private SourceDataLine speakers;

    // Control de gRPC Stream
    private StreamObserver<AudioChunk> requestObserver; // Para enviar audio al servidor
    private boolean grpcStreamActive = false;

    // Control de estados individuales
    private volatile boolean micActive = false;
    private volatile boolean speakersActive = false;

    private Thread micCaptureThread; // Hilo para capturar del micrófono

    public AudioStreamer(ChatServiceGrpc.ChatServiceStub asyncStub, String sender, String roomId) {
        this.asyncStub = asyncStub;
        this.sender = sender;
        this.roomId = roomId;
        this.audioFormat = getAudioFormat();
    }

    private AudioFormat getAudioFormat() {
        return new AudioFormat(44100, 16, 1, true, false); // 44.1kHz, 16bit, Mono, Signed, Little-endian
    }

    // Helper para imprimir mensajes Y redibujar el prompt
    private void printMessage(String message) {
        String ansiClearLine = "\r\u001b[2K";
        String prompt = "[" + LocalDateTime.now().format(TIME_FORMATTER) + "] Tú: ";
        System.out.print(ansiClearLine + message + "\n" + prompt);
        System.out.flush();
    }


    /**
     * Establece la conexión bidireccional gRPC para audio.
     * Esto DEBE llamarse antes de intentar usar el micrófono o los altavoces.
     */
    public void startAudioConnection() {
        if (grpcStreamActive) {
            printMessage("La conexión de audio gRPC ya está activa.");
            return;
        }

        // Observer para recibir audio del servidor y reproducirlo
        StreamObserver<AudioChunk> responseObserver = new StreamObserver<AudioChunk>() {
            @Override
            public void onNext(AudioChunk value) {
                if (speakersActive && speakers != null && speakers.isOpen()) {
                    byte[] audioData = value.getData().toByteArray();
                    speakers.write(audioData, 0, audioData.length);
                }
            }

            @Override
            public void onError(Throwable t) {
                String errorDetails;
                if (t instanceof io.grpc.StatusRuntimeException) {
                    errorDetails = ((io.grpc.StatusRuntimeException) t).getStatus().toString();
                } else {
                    errorDetails = t.getMessage();
                }
                System.err.println("Error en el stream de audio gRPC: " + errorDetails); // Mantener System.err para errores
                stopAudioConnection(); // Asegurarse de cerrar todo
            }

            @Override
            public void onCompleted() {
                printMessage("Recepción de audio gRPC finalizada.");
                stopAudioConnection(); // Asegurarse de cerrar todo
            }
        };

        // Adjuntar metadatos al contexto de la llamada gRPC
        ClientInterceptor headerInterceptor = new ClientInterceptor() {
            @Override
            public <ReqT, RespT> io.grpc.ClientCall<ReqT, RespT> interceptCall(
                    MethodDescriptor<ReqT, RespT> method,
                    io.grpc.CallOptions callOptions,
                    io.grpc.Channel next) {
                return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
                    @Override
                    public void start(ClientCall.Listener<RespT> responseListener, Metadata headersToAdd) {
                        headersToAdd.put(Metadata.Key.of("room-id", Metadata.ASCII_STRING_MARSHALLER), roomId);
                        headersToAdd.put(Metadata.Key.of("sender", Metadata.ASCII_STRING_MARSHALLER), sender);
                        super.start(new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(responseListener) {
                            // No es necesario añadir lógica aquí para el POC
                        }, headersToAdd);
                    }
                };
            }
        };

        requestObserver = asyncStub.withInterceptors(headerInterceptor).streamAudio(responseObserver);
        grpcStreamActive = true;
        printMessage("Conexión de audio gRPC establecida.");
    }

    /**
     * Cierra la conexión bidireccional gRPC para audio.
     * Esto detendrá tanto el envío como la recepción.
     */
    public void stopAudioConnection() {
        if (!grpcStreamActive) {
            return;
        }
        grpcStreamActive = false;
        stopMic(); // Asegurarse de detener el micrófono si está activo
        stopSpeakers(); // Asegurarse de detener los altavoces si están activos

        if (requestObserver != null) {
            requestObserver.onCompleted();
        }
        printMessage("Conexión de audio gRPC cerrada.");
    }

    /**
     * Inicia la captura de audio del micrófono y el envío al servidor.
     */
    public void startMic() {
        if (micActive) {
            printMessage("Micrófono ya activo.");
            return;
        }
        if (!grpcStreamActive) {
            System.err.println("Primero debes establecer la conexión gRPC de audio (/mic on o /listen on)."); // Mantener System.err
            return;
        }

        try {
            DataLine.Info info = new DataLine.Info(TargetDataLine.class, audioFormat);
            microphone = (TargetDataLine) AudioSystem.getLine(info);
            microphone.open(audioFormat);
            microphone.start();

            micActive = true;
            printMessage("Micrófono activado. Transmitiendo voz...");

            micCaptureThread = new Thread(() -> {
                byte[] buffer = new byte[1024]; // Tamaño del buffer
                while (micActive && requestObserver != null) {
                    int bytesRead = microphone.read(buffer, 0, buffer.length);
                    if (bytesRead > 0) {
                        AudioChunk chunk = AudioChunk.newBuilder()
                                .setData(ByteString.copyFrom(buffer, 0, bytesRead))
                                .build();
                        try {
                            requestObserver.onNext(chunk);
                        } catch (Exception e) {
                            System.err.println("Error al enviar chunk de audio: " + e.getMessage()); // Mantener System.err
                            micActive = false; // Detener el hilo si falla el envío
                        }
                    } else {
                        // System.err.println("DEBUG: Micrófono no leyó bytes, esperando..."); // DEBUG LOG - Eliminado
                        try {
                            Thread.sleep(100); // Esperar un poco si no hay bytes para evitar bucle ocupado
                        } catch (InterruptedException ie) {
                            Thread.currentThread().interrupt();
                        }
                    }
                }
                microphone.close();
                printMessage("Captura de micrófono detenida.");
            });
            micCaptureThread.setDaemon(true);
            micCaptureThread.start();

        } catch (LineUnavailableException e) {
            System.err.println("Error al acceder al micrófono: " + e.getMessage()); // Mantener System.err
            micActive = false;
        }
    }

    /**
     * Detiene la captura de audio del micrófono.
     */
    public void stopMic() {
        if (!micActive) {
            return;
        }
        micActive = false;
        if (micCaptureThread != null) {
            micCaptureThread.interrupt(); // Interrumpir el hilo si está esperando
            try {
                micCaptureThread.join(500); // Esperar un poco a que termine
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
        if (microphone != null && microphone.isOpen()) {
            microphone.stop();
            microphone.close();
        }
        printMessage("Micrófono detenido.");
    }

    /**
     * Inicia la reproducción de audio recibido por los altavoces.
     */
    public void startSpeakers() {
        if (speakersActive) {
            printMessage("Altavoces ya activos para reproducción.");
            return;
        }
        if (!grpcStreamActive) {
            System.err.println("Primero debes establecer la conexión gRPC de audio (/mic on o /listen on)."); // Mantener System.err
            return;
        }

        try {
            DataLine.Info info = new DataLine.Info(SourceDataLine.class, audioFormat);
            speakers = (SourceDataLine) AudioSystem.getLine(info);
            speakers.open(audioFormat);
            speakers.start();
            speakersActive = true;
            printMessage("Altavoces activados para reproducción de audio.");
        } catch (LineUnavailableException e) {
            System.err.println("Error al acceder a los altavoces: " + e.getMessage()); // Mantener System.err
            speakersActive = false;
        }
    }

    /**
     * Detiene la reproducción de audio por los altavoces.
     */
    public void stopSpeakers() {
        if (!speakersActive) {
            return;
        }
        speakersActive = false;
        if (speakers != null && speakers.isOpen()) {
            speakers.drain();
            speakers.close();
        }
        printMessage("Altavoces detenidos.");
    }

    public boolean isMicActive() {
        return micActive;
    }

    public boolean isSpeakersActive() {
        return speakersActive;
    }

    public boolean isGrpcStreamActive() {
        return grpcStreamActive;
    }
}